daemons/controld/controld_te_callbacks.c

```diff
@@ -902,30 +904,44 @@ cib_action_updated(xmlNode * msg, int call_id, int rc, xmlNode * output, void *u
     }
 }
 
+/*!
+ * \brief Handle a timeout in node-to-node communication
+ *
+ * \param[in] data  Pointer to action timer
+ *
+ * \return FALSE (indicating that source should be not be re-added)
+ */
 gboolean
 action_timer_callback(gpointer data)
 {
     crm_action_timer_t *timer = NULL;
+    const char *task = NULL;
+    const char *on_node = NULL;
+    const char *via_node = NULL;
 
     CRM_CHECK(data != NULL, return FALSE);
 
     timer = (crm_action_timer_t *) data;
     stop_te_timer(timer);
 
-    crm_warn("Timer popped (timeout=%d, abort_level=%d, complete=%s)",
-             timer->timeout,
-             transition_graph->abort_priority, transition_graph->complete ? "true" : "false");
-
     CRM_CHECK(timer->action != NULL, return FALSE);
 
-    if (transition_graph->complete) {
-        crm_warn("Ignoring timeout while not in transition");
+    task = crm_element_value(timer->action->xml, XML_LRM_ATTR_TASK);
+    on_node = crm_element_value(timer->action->xml, XML_LRM_ATTR_TARGET);
+    via_node = crm_element_value(timer->action->xml, XML_LRM_ATTR_ROUTER_NODE);
 
+    if (transition_graph->complete) {
+        crm_notice("Node %s did not send %s result (via %s) within %dms "
+                   "(ignoring because transition not in progress)",
+                   (on_node? on_node : ""), (task? task : "unknown action"),
+                   (via_node? via_node : "controller"), timer->timeout);
     } else {
         /* fail the action */
-        gboolean send_update = TRUE;
-        const char *task = crm_element_value(timer->action->xml, XML_LRM_ATTR_TASK);
 
+        crm_err("Node %s did not send %s result (via %s) within %dms "
+                "(action timeout plus cluster-delay)",
+                (on_node? on_node : ""), (task? task : "unknown action"),
+                (via_node? via_node : "controller"), timer->timeout);
```

Comment:
`Node<two spaces>did not send` may look weird.

Indeed, the deeper and more serious problem alredy popping up
(ad nauseam) is that we have no level of declarative assurance
of how the messages may look like -- all the information about
optionality etc. is interspersed over the codebase without
possibility to make any firmer assumptions about the structure,
yielding superfluous checks over and over.  For instance, here,
synthetic operations (`inject_op`) will indeed have no `node`
defined.  That were formalized/structured abiding messages would
help, and for local exchanges, binary structures would be just
fine, see https://bugs.clusterlabs.org/show_bug.cgi?id=5376#c3.

Reply:
While the double space could be distracting, it's not particularly significant, and that case should never happen in practice (inject_op() doesn't actually go through the controller and so will never time out) -- it's just a failsafe in case of a bug elsewhere.

Reply:
`inject_op()` was just an example, not sure if it's a good use of
energe to do the comprehensive searches in cases like that (that's
why I wished there was a clear structure and thanks to that additive
assurances like stated -- no cumulative synergy is possible where
"everything is allowed")